From ee8fe1470c74d6f324e5fa6fb9fc820d27a0d4a4 Mon Sep 17 00:00:00 2001
From: Arindam Bhattacharjee <abhattacharjee717@gmail.com>
Date: Wed, 19 Nov 2025 19:35:38 +0000
Subject: [PATCH 2/2] Introduce `Force LTE_CA` override on per subscription
 basis [3/3]

Patches required in:
- frameworks/base [1/3]
- frameworks/opt/telephony [2/3]
- packages/apps/settings [3/3]

Workflow after this patch:
---------------------------------------------------------------------------------------------
Hardware CA | Force LTE_CA Toggle | Code Path                                  | Icon Shows |
------------|---------------------|--------------------------------------------|------------|
No          | Unset (null)        | Returns false                              | LTE        |
No          | ON (true)           | Software-based Forced LTE-CA (mForceLteCA) | LTE+       |
No          | OFF (false)         | Returns mForceLteCA (false)                | LTE        |
Yes         | Unset (null)        | Returns false                              | LTE        |
Yes         | ON (true)           | Returns mForceLteCA (true)                 | LTE+       |
Yes         | OFF (false)         | Returns mForceLteCA (false)                | LTE        |
---------------------------------------------------------------------------------------------

Change-Id: Ife6892371d68d03a29427e8ce753d2edba89b3d5
Signed-off-by: Arindam Bhattacharjee <abhattacharjee717@gmail.com>
Signed-off-by: Pranav Vashi <neobuddy89@gmail.com>
---
 res/xml/mobile_network_settings.xml           |  5 +++
 .../telephony/MobileNetworkSettings.java      | 31 +++++++++++++++++++
 2 files changed, 36 insertions(+)

diff --git a/res/xml/mobile_network_settings.xml b/res/xml/mobile_network_settings.xml
index a317cd04cd77..e924a7a2c4a3 100644
--- a/res/xml/mobile_network_settings.xml
+++ b/res/xml/mobile_network_settings.xml
@@ -283,6 +283,11 @@
                 android:key="choose_network_key"
                 android:title="@string/choose_network_title"
                 settings:controller="com.android.settings.network.telephony.gsm.OpenNetworkSelectPagePreferenceController"/>
+
+            <SwitchPreferenceCompat
+                android:key="force_lte_ca"
+                android:title="@string/force_lte_ca_title"
+                android:summary="@string/force_lte_ca_summary" />
         </PreferenceCategory>
 
         <com.android.settingslib.widget.UntitledPreferenceCategory
diff --git a/src/com/android/settings/network/telephony/MobileNetworkSettings.java b/src/com/android/settings/network/telephony/MobileNetworkSettings.java
index 8fcb3b526ebe..3b77f7000d81 100644
--- a/src/com/android/settings/network/telephony/MobileNetworkSettings.java
+++ b/src/com/android/settings/network/telephony/MobileNetworkSettings.java
@@ -42,6 +42,7 @@ import androidx.annotation.VisibleForTesting;
 import androidx.lifecycle.LifecycleOwner;
 import androidx.lifecycle.ViewModelProvider;
 import androidx.preference.Preference;
+import androidx.preference.SwitchPreferenceCompat;
 
 import com.android.settings.R;
 import com.android.settings.Settings.MobileNetworkActivity;
@@ -96,6 +97,9 @@ public class MobileNetworkSettings extends AbstractMobileNetworkSettings impleme
     private static final String BUTTON_CDMA_SYSTEM_SELECT_KEY = "cdma_system_select_key";
     private static final String BUTTON_CDMA_SUBSCRIPTION_KEY = "cdma_subscription_key";
 
+    private static final String KEY_FORCE_LTE_CA = "force_lte_ca";
+    private SwitchPreferenceCompat mForceLteCAPreference;
+
     private final ExecutorService mExecutor = Executors.newSingleThreadExecutor();
 
     private TelephonyManager mTelephonyManager;
@@ -354,6 +358,19 @@ public class MobileNetworkSettings extends AbstractMobileNetworkSettings impleme
                 setTelephonyAvailabilityStatus(getPreferenceControllersAsList());
 
         super.onCreate(icicle);
+
+        mForceLteCAPreference = findPreference(KEY_FORCE_LTE_CA);
+        mForceLteCAPreference.setOnPreferenceChangeListener((preference, newValue) -> {
+            boolean enabled = (Boolean) newValue;
+            int phoneId = SubscriptionManager.getPhoneId(mSubId);
+            Settings.Global.putInt(
+                getContext().getContentResolver(),
+                Settings.Global.FORCE_LTE_CA + "_" + phoneId,
+                enabled ? 1 : 0
+            );
+            return true;
+        });
+
         if (isUiRestricted()) {
             Log.d(LOG_TAG, "Mobile network page is disallowed.");
             finish();
@@ -394,6 +411,9 @@ public class MobileNetworkSettings extends AbstractMobileNetworkSettings impleme
         super.onResume();
         mMobileNetworkRepository.addRegister(this, this, mSubId);
         mMobileNetworkRepository.updateEntity();
+
+        updateForceLteCAState();
+
         // TODO: remove log after fixing b/182326102
         Log.d(LOG_TAG, "onResume() subId=" + mSubId);
 
@@ -402,6 +422,17 @@ public class MobileNetworkSettings extends AbstractMobileNetworkSettings impleme
         getContext().registerReceiver(mBrocastReceiver, intentFilter, Context.RECEIVER_EXPORTED);
     }
 
+    private void updateForceLteCAState() {
+        if (mForceLteCAPreference != null) {
+            int phoneId = SubscriptionManager.getPhoneId(mSubId);
+            boolean enabled = Settings.Global.getInt(
+                    getContext().getContentResolver(),
+                    Settings.Global.FORCE_LTE_CA + "_" + phoneId,
+                    0) == 1;
+            mForceLteCAPreference.setChecked(enabled);
+        }
+    }
+
     private void onSubscriptionDetailChanged() {
         final SubscriptionInfoEntity subscriptionInfoEntity = mSubscriptionInfoEntity;
         if (subscriptionInfoEntity == null) {
-- 
2.43.0

